File: repos_estudiantes_TARQUI_tercera_entrega/ROJAS_ALCON_VICTOR_ALEJANDRO/generecidad/ej3/ej3.java
import java.util.ArrayList;

class Catalogo<T> {
    private ArrayList<T> elementos = new ArrayList<>();

    public void agregar(T item) {
        elementos.add(item);
    }

    public boolean buscar(T item) {
        return elementos.contains(item);
    }

    public void mostrarTodo() {
        for (T item : elementos) {
            System.out.println(item);
        }
    }
}

class Libro {
    String titulo;

    Libro(String titulo) {
        this.titulo = titulo;
    }

    @Override
    public String toString() {
        return "Libro: " + titulo;
    }

    @Override
    public boolean equals(Object o) {
        if (o instanceof Libro) {
            return this.titulo.equals(((Libro) o).titulo);
        }
        return false;
    }
}

class Producto {
    String nombre;

    Producto(String nombre) {
        this.nombre = nombre;
    }

    @Override
    public String toString() {
        return "Producto: " + nombre;
    }

    @Override
    public boolean equals(Object o) {
        if (o instanceof Producto) {
            return this.nombre.equals(((Producto) o).nombre);
        }
        return false;
    }
}

public class Main {
    public static void main(String[] args) {
        Catalogo<Libro> catalogoLibros = new Catalogo<>();
        catalogoLibros.agregar(new Libro("El Principito"));
        catalogoLibros.agregar(new Libro("1984"));
        System.out.println("Buscar '1984': " + catalogoLibros.buscar(new Libro("1984")));
        catalogoLibros.mostrarTodo();

        Catalogo<Producto> catalogoProductos = new Catalogo<>();
        catalogoProductos.agregar(new Producto("Laptop"));
        catalogoProductos.agregar(new Producto("Mouse"));
        System.out.println("Buscar 'Teclado': " + catalogoProductos.buscar(new Producto("Teclado")));
        catalogoProductos.mostrarTodo();
    }
}

File: repos_estudiantes_TARQUI_tercera_entrega/ROJAS_ALCON_VICTOR_ALEJANDRO/generecidad/ej3/ej3.py
from typing import Generic, TypeVar, List

T = TypeVar('T')

class Catalogo(Generic[T]):
    def __init__(self):
        self.elementos: List[T] = []

    def agregar(self, item: T):
        self.elementos.append(item)

    def buscar(self, item: T) -> bool:
        return item in self.elementos

    def mostrar_todo(self):
        for item in self.elementos:
            print(item)

class Libro:
    def __init__(self, titulo: str):
        self.titulo = titulo

    def __eq__(self, other):
        return isinstance(other, Libro) and self.titulo == other.titulo

    def __str__(self):
        return f"Libro: {self.titulo}"

class Producto:
    def __init__(self, nombre: str):
        self.nombre = nombre

    def __eq__(self, other):
        return isinstance(other, Producto) and self.nombre == other.nombre

    def __str__(self):
        return f"Producto: {self.nombre}"

catalogo_libros = Catalogo[Libro]()
catalogo_libros.agregar(Libro("El Principito"))
catalogo_libros.agregar(Libro("1984"))
print("Buscar '1984':", catalogo_libros.buscar(Libro("1984")))
catalogo_libros.mostrar_todo()

catalogo_productos = Catalogo[Producto]()
catalogo_productos.agregar(Producto("Laptop"))
catalogo_productos.agregar(Producto("Mouse"))
print("Buscar 'Teclado':", catalogo_productos.buscar(Producto("Teclado")))
catalogo_productos.mostrar_todo()
